File: backend/app/schemas/releases.py
from datetime import datetime, timezone
from typing import Literal
from uuid import UUID
from pydantic import BaseModel, ConfigDict, Field, field_validator


class Input(BaseModel):
    model_config = ConfigDict(extra="forbid")


class Login(Input):
    token: str = Field(min_length=1, max_length=200)


class SourceImport(Input):
    url: str = Field(max_length=2000)
    text: str | None = Field(default=None, min_length=20, max_length=200_000)
    title: str | None = Field(default=None, max_length=500)
    publisher: str | None = Field(default=None, max_length=200)
    published_at: datetime | None = None
    fallback_published_at: datetime | None = None
    event_name: str | None = Field(default=None, max_length=200)
    season: int = Field(default=2026, ge=2026, le=2026)


class CandidateReview(Input):
    legacy_upgrade_id: UUID | None = None
    team_key: Literal["ferrari", "mercedes"] | None = None
    component: str | None = Field(default=None, max_length=60)
    event_name: str | None = Field(default=None, max_length=200)
    observed_at: datetime | None = None
    page: int | None = Field(default=None, ge=1, le=250)
    summary: str | None = Field(default=None, min_length=5, max_length=2000)
    supporting_passage: str | None = Field(default=None, min_length=5, max_length=4000)
    evidence_status: (
        Literal["unverified", "reported", "confirmed", "conflicting"] | None
    ) = None
    representation: Literal["annotation_only", "modeled"] | None = None
    review_notes: str = Field(default="", max_length=4000)
    status: Literal["draft", "approved", "rejected"] = "draft"

    @field_validator(
        "summary", "supporting_passage", "evidence_status", "representation"
    )
    @classmethod
    def required_when_supplied(cls, value):
        if value is None:
            raise ValueError("This field cannot be cleared.")
        return value


class RevisionInput(Input):
    component: str
    parameters: dict[str, float] = Field(default_factory=dict)
    source_ids: list[UUID] = Field(min_length=1, max_length=30)
    uncertainty: str = Field(min_length=10, max_length=4000)


class VersionInput(Input):
    team_key: Literal["ferrari", "mercedes"]
    season: Literal[2026] = 2026
    label: str = Field(min_length=3, max_length=200)
    configuration_event: str = Field(min_length=3, max_length=200)
    configuration_kind: Literal[
        "baseline",
        "evolution",
        "circuit_specific",
        "reversion",
        "no_change",
        "reconstruction",
    ] = "evolution"
    reverts_to_id: UUID | None = None
    as_of: datetime
    evidence_cutoff: datetime
    parent_id: UUID | None = None
    candidate_ids: list[UUID] = Field(default_factory=list, max_length=200)
    revisions: list[RevisionInput] = Field(default_factory=list, max_length=20)
    notes: str = Field(default="", max_length=4000)

    @field_validator("as_of", "evidence_cutoff")
    @classmethod
    def aware_date(cls, value):
        return (
            value.replace(tzinfo=timezone.utc)
            if value.tzinfo is None
            else value.astimezone(timezone.utc)
        )


class VisualReview(Input):
    front: bool
    side: bool
    rear: bool
    three_quarter: bool
    reference_urls: list[str] = Field(min_length=1, max_length=20)
    notes: str = Field(min_length=20, max_length=4000)


class Rollback(Input):
    version_id: UUID
    reason: str = Field(min_length=5, max_length=1000)

File: frontend/src/lib/releases.ts
export const TEAM_NAMES = {
  ferrari: "Ferrari",
  mercedes: "Mercedes",
  mclaren: "McLaren",
  red_bull: "Red Bull Racing",
  aston_martin: "Aston Martin",
  alpine: "Alpine",
  williams: "Williams",
  haas: "Haas",
  racing_bulls: "Racing Bulls",
  audi: "Audi",
  cadillac: "Cadillac",
} as const;
export type TeamKey = keyof typeof TEAM_NAMES;
export const TEAM_KEYS = Object.keys(TEAM_NAMES) as TeamKey[];
export type ViewName = "three_quarter" | "front" | "side" | "rear";
export interface Source {
  id: string;
  url: string;
  title: string;
  publisher: string;
  published_at: string | null;
  retrieved_at: string;
  source_type: string;
  image_urls: string[];
  rights: string;
}
export interface Candidate {
  id: string;
  source_id: string;
  team_key: TeamKey | null;
  season: number;
  component: string | null;
  event_name: string | null;
  observed_at: string | null;
  summary: string;
  supporting_passage: string;
  page: number | null;
  evidence_status: "unverified" | "reported" | "confirmed" | "conflicting";
  representation: "annotation_only" | "modeled";
  status: "draft" | "approved" | "rejected";
  review_notes: string;
  source?: Source;
}
export interface Asset {
  filename: string;
  url: string;
  sha256: string;
  bytes: number;
}
export interface ComponentState {
  revision_id: string;
  label: string;
  anchor: [number, number, number];
  parameters: Record<string, number>;
  source_ids: string[];
  uncertainty: string;
  changed: boolean;
}
export interface CarVersion {
  id: string;
  team_key: TeamKey;
  season: number;
  label: string;
  configuration_event: string;
  configuration_kind: string;
  reverts_to_id?: string | null;
  as_of: string;
  evidence_cutoff: string;
  status: "draft" | "building" | "ready" | "published" | "failed";
  parent_id: string | null;
  published_at: string | null;
  is_current: boolean;
  notes: string;
  component_revisions: Record<string, string>;
  visual_review: Partial<Record<ViewName, boolean>> & {
    notes?: string;
    reference_urls?: string[];
    reviewed_at?: string;
  };
  manifest: {
    components: Record<string, ComponentState>;
    sources: Source[];
    changes: Candidate[];
    assets: Record<string, Asset>;
    component_hashes?: Record<string, string>;
    no_new_modeled_change: boolean;
    reconstruction_notice: string;
    validation?: { triangles: number; glb_bytes: number; textures: number };
  };
}
export interface Catalog {
  teams: Record<
    TeamKey,
    {
      name: string;
      car_name: string;
      accent: string;
      baseline_date: string;
      notes: string;
      source_url: string;
    }
  >;
  components: Record<
    string,
    {
      label: string;
      anchor: [number, number, number];
      parameters: Record<string, [number, number]>;
    }
  >;
  season: number;
}
export interface Job {
  id: string;
  version_id: string | null;
  kind: string;
  status: string;
  attempts: number;
  error: string | null;
  result: Record<string, unknown>;
  created_at: string;
  started_at: string | null;
  finished_at: string | null;
}
export interface Dashboard {
  versions: CarVersion[];
  candidates: Candidate[];
  jobs: Job[];
  sources: Source[];
  audit: { action: string; target_id: string; created_at: string }[];
}

export async function request<T>(
  path: string,
  init: RequestInit = {},
): Promise<T> {
  const response = await fetch(`/api/v1${path}`, {
    ...init,
    credentials: "same-origin",
    cache: "no-store",
    headers: {
      "Content-Type": "application/json",
      "X-F1-Review": "1",
      ...init.headers,
    },
  });
  if (!response.ok) {
    let detail = `Request failed (${response.status})`;
    try {
      const body = await response.json();
      detail =
        typeof body.detail === "string"
          ? body.detail
          : JSON.stringify(body.detail);
    } catch {
      /* proxy may return HTML */
    }
    throw new Error(detail);
  }
  return response.json();
}
export const mutate = <T>(path: string, body: unknown = {}, method = "POST") =>
  request<T>(path, { method, body: JSON.stringify(body) });
export const dateLabel = (date: string) =>
  new Date(date).toLocaleDateString("en-GB", {
    day: "numeric",
    month: "short",
    year: "numeric",
    timeZone: "UTC",
  });
